### ISSUE TITLE
Add a few unittest2 assertions to TestCase

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6350]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)| @Julian reported|
|-|-|
|Trac ID|trac#6350|
|Type|enhancement|
|Created|2013-03-03 21:16:00Z|
|Branch|https://github.com/twisted/twisted/tree/assertIs-6350|

This begins to add some of unittest2's assertions to TestCase. Some of these are needed (assertIs and assertIsNot) because they will be replacing assertIdentical and assertNotIdentical.

Others are just plain dumb (assertIsNone) though unittest2 has them, and eventually it would be nice to have full unittest2 compatibility (once again, unittest2 is the name for pyunit in 2.7+). So these are here for now to help that move forward. Eventually they will be removed and only defined on Python 2.6.

Attachments:

* [assertIs.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/93a/93acdb394cba40fd309de04612f854386b65a388/da3ee46a9a370eb3e0d95f916265ff78d97354b8.patch) (5177 bytes) - added by Julian on 2013-03-03 21:19:35Z - 
* [assertIs-v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/93a/93acdb394cba40fd309de04612f854386b65a388/99314862a00acda1ab13296dfcb3fb9630767f99.patch) (2670 bytes) - added by Julian on 2013-03-04 14:05:25Z - 
* [assertIs-v3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/93a/93acdb394cba40fd309de04612f854386b65a388/d5cd3bc03e30ffc28f86bee5f2c0593162db02a9.patch) (2874 bytes) - added by Julian on 2013-03-22 04:07:41Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6350 6350
type__enhancement enhancement
reporter__Julian Julian
priority__normal normal
milestone__ 
branch__branches_assertIs_6350 branches/assertIs-6350
branch_author__tomprince tomprince
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1362345360000000 1362345360000000
changetime__1365183496000000 1365183496000000
version__None None
owner__tom_prince tom.prince
cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

> eventually it would be nice to have full unittest2 compatibility

I'm not so sure about this.  Keep in mind that trial `TestCase` is a subclass of `unittest.TestCase`.  It inherits all the methods `unittest.TestCase` defines.  It is hardly a surprise that `unittest.TestCase` on Python 2.6 does not have the new features of `unittest.TestCase` which were added in Python 2.7.  Generally speaking, this is not a problem.  Developers writing code for Python 2.6 are aware that they will be getting Python 2.6 functionality, not Python 2.7 functionality.

If there are specific cases where some neat Python 2.7 `unittest` functionality would actually help implement some tests for Twisted, by all means extend trial to provide this functionality on Python 2.6 as well.  However, functionality should be driven by real use cases (because only through real use does code provide actual value).  In the abstract (that is, without real use cases), "full compatibility" is meaningless (only `unittest2` is _fully_ compatible with `unittest2`) and not obviously desirable.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Other areas of trial which could benefit greatly from attention (in place of "full compatibility" with `unittest2`):

  * [#7201](https://github.com/twisted/twisted/issues/7201)
  * [#2365](https://github.com/twisted/twisted/issues/2365)
  * [#3178](https://github.com/twisted/twisted/issues/3178)

And if you want to focus on stdlib interoperability, then [#6345](https://github.com/twisted/twisted/issues/6345) would help solve at least one person's real-world issue.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_4"></a>|@Julian commented|
|-|-|

Well. When I say "full compatibility", it'd be nice at least if trial were able to take advantage of unittest2 if present. E.g. if most of the time I write 2.7+ code, when I want to use a trial TestCase I'm forced to use 2.6 style assertions.

If I have unittest2 installed though, I'd like trial to pick it up and inherit from unittest2.TestCase rather than unittest, same as if I were writing non-twisted test code, where I could use 2.7 style assertions even though I was supporting 2.6 too.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_5"></a>|@Julian commented|
|-|-|

Though assertIsNone is essentially totally pointless given that even in pyunit it doesn't give any better of an error than assertIs(foo, None). So the real world use case here is that as things stand, when I write non-twisted code, I'm able to write 2.7+ style assertions and then have them work on 2.6.

I can't do so when I write twisted code, and it'd be nice if I could.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

> So the real world use case here is that as things stand, when I write non-twisted code, I'm able to write 2.7+ style assertions and then have them work on 2.6.

This is the problem with making `TestCase` the center of extensibility for making assertions.  `unittest2` is just one example of a divergence in the class hierarchy that can't be resolved except by duplicating lots of code all over the place.

It seems to me that adding [testtools](https://pypi.python.org/pypi/testtools) style matcher support would solve this problem once and for all.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_7"></a>|@Julian commented|
|-|-|

OK well, we don't especially need to have this conversation now I guess. To get us unblocked on [#6220](https://github.com/twisted/twisted/issues/6220) here's just the addition of assertIs and assertIsNot, which I hope you at least agree on (that we should use the unittest2 names for assertions we already have but under different names).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_8"></a>|@tomprince set owner to @Julian|
|-|-|

1.The tests don't check that the assertions use `is` instead of `==`:
```python
(object() == object()) is False
```
 2. The test docstrings should document what the test is doing, and what result it expects from that (see [#6301](https://github.com/twisted/twisted/issues/6301)). We want to document the intent of the test, so that a reviewer, or future reader can see if the test is testing what is expected.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_9"></a>|@Julian removed owner|
|-|-|

Sigh. These were copied out of the stdlib test suite :).

Fixed, see latest patch. Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @Julian|
|-|-|

> Sigh. These were copied out of the stdlib test suite :).

Um.  I am quite concerned about this.  You cannot copy code out of other projects and contribute it to Twisted without annotating its origin.  Doing so is potentially a violation of the Python license and potentially makes distributing Twisted itself illegal.

Taking this ticket out of review until this issue is addressed definitively.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_11"></a>|@Julian commented|
|-|-|

There is no directly copied code in here and there never was.

The assertion in the test (an assertEqual on two things that happen to be objects()) was the same as what the stdlib happens to be doing, so all I meant was that the stdlib didn't care about specifically going out of its way to complicate things.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)|@Julian removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_13"></a>|@tomprince commented|
|-|-|

(In [37931]) Branching to assertIs-6350.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_14"></a>|@tomprince commented|
|-|-|

(In [37932]) Apply assertIs-v3.patch from Julian.

Refs: [#6350](https://github.com/twisted/twisted/issues/6350)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_15"></a>|@tomprince set owner to @tomprince|
|-|-|

This looks good. I've added a topfile.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_16"></a>|@tomprince set status to `closed`|
|-|-|

(In [37934]) Merge assertIs-6350: Add assertIs and assertIsNot assertions.

Author: Julian
Reviewers: tom.prince
Fixes: [#6350](https://github.com/twisted/twisted/issues/6350)

These are the names used in python 2.7 and python 3.
Adding these is part of picking a single spelling for
each assertion.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_17"></a>|@Julian commented|
|-|-|

Great, thanks as usual Tom. See you on the flipside of [#6220](https://github.com/twisted/twisted/issues/6220) and friends.
